### PR TITLE
Fixed bug while adding agents with 'any' as IP with authd

### LINKED ISSUE #2497: Cannot register more than one agent as "any" using agent-auth if the tag <use_source_ip> is set to yes
Hello team,

I tried to register agents through NAT using the `ossec-authd` daemon and configuring the IP "any" on the agent side using the command `agent-auth -I any`. 

This way, when I registered more of 1 agent using that command, the manager was always overwriting the previous one. The manager only registered one agent with "any" IP.

I had set the tag`use_source_ip` in the manager to no, to register the agent as "any" and not use the option `-I any`.

Is this the expected behavior?

Regards,

Miguel Casares

### PR DESCRIPTION
**Pull request by DaveVG1**
This fixes #2497 
- Added condition to check if a new IP is duplicated while adding a new agent with `authd`: if the IP is `any`, we don't have to check if it's duplicated.
- Changed maximum length of IP expected from `authd` (with the parameter `-I`) in order to match every possible IP (IPs larger than 15 characteres were being truncated, so we couldn't add the netmask).
- Fixed bug occured while adding an agent with netmask in its IP via `authd`. `agent-info` and `ruleset` files were being deleted after initializating the agent (it was stored in the database with the netmask, but it was being search later without it).

### CODE REVIEW

**Comment by DaveVG1** [comment]
Testing
-----------------
- Added several agents with `any` as IP.
- Added agent with netmask (won't delete `agent-info` file anymore).

Known issues
----
- If we register an agent with netmask (`192.168.1.200`, e.g. ) and register another with the same IP but without netmask (in that specific order), the first one will be overwritten by the second.

**Comment by sempervictus** [comment]
Thank you. I think this explains the Windows workstation weirdness we've been seeing.

File: src/wazuh_modules/wm_database.c [hunk 1]
```diff
@@ -848,7 +848,7 @@ int wm_sync_file(const char *dirname, const char *fname) {
         if (id_agent) {
             switch (wm_extract_agent(fname, name, addr, &is_registry)) {
             case 0:
-                if ((id_agent = wdb_find_agent(name, addr)) < 0) {
+                if ((id_agent = wdb_find_agent(name, NULL)) < 0) {
```
Comment (nicolaspapp) [inline src/wazuh_modules/wm_database.c]: Is there any particular reason why we need this change? The SQL_FIND_AGENT query seems to work even when the address is '**_any'_**

File: src/wazuh_db/wdb_agent.c [hunk 2]
```diff
@@ -430,14 +431,24 @@ int wdb_find_agent(const char *name, const char *ip) {
     if (wdb_open_global() < 0)
         return -1;
 
-    if (wdb_prepare(wdb_global, SQL_FIND_AGENT, -1, &stmt, NULL)) {
-        mdebug1("SQLite: %s", sqlite3_errmsg(wdb_global));
-        wdb_close_global();
-        return -1;
+    if (ip) {
```
Comment (nicolaspapp) [inline src/wazuh_db/wdb_agent.c]: if we change ` wdb_find_agent(name, NULL))` for ` wdb_find_agent(name, addr))` the distintion does not seem necessary anymore

**Comment by nicolaspapp** [comment]
## Test Scenarios

### 1 -  2 Hosts using `agent-auth -I any` and `<use_source_ip>no</use_source_ip>`:

```
002 localhost.localdomain any 6a49da84ab49e62dcf341acf967064871e3e11914e0b89678a7cdccf3ea5d4c1
003 hostname.domain.com any 2f731ba94ce784456287d42714fb629dbafd52e9154c908ee1ac28fd814353b1
```

### 2 -  2 Hosts using `agent-auth -I any` and `<use_source_ip>yes</use_source_ip>`:

```
005 hostname2.domain.com 192.168.56.4 e900d578ace6a3d0aa0b53b67d6c13f1ea72902b9a679acbdf51c272563b9ad2
006 hostname.domain.com 192.168.56.3 31c50536059befeac346c23cd5fd36d974f22b2b0f83cb7cb80ebd20c8eb1dc1
```

### 3 - Same host, 2 registrations with  `agent-auth -I any` and `<use_source_ip>no</use_source_ip>`:

```
001 localhost.localdomain any {key 1} -> deleted after 2nd registration
002 localhost.localdomain any {key 2}
```

**Comment by wazuhci** [comment]
The stress test has been launched. The charts will be posted when they are ready.
I will use the following configuration:
### General
- Commit: d40cb9c
- Packages: 
  - [manager 3.12.0 0.branchfixauthdipany.commitd40cb9c rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-manager-3.12.0-0.branchfixauthdipany.commitd40cb9c.x86_64.rpm)
  - [agent 3.12.0 0.branchfixauthdipany.commitd40cb9c rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-agent-3.12.0-0.branchfixauthdipany.commitd40cb9c.x86_64.rpm)
  - [manager 3.12.0 0.branchfixauthdipany.commitd40cb9c deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-manager_3.12.0-0.branchfixauthdipany.commitd40cb9c_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commitd40cb9c deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-agent_3.12.0-0.branchfixauthdipany.commitd40cb9c_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commitd40cb9c msi](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/windows/wazuh-agent-3.12.0-0.branchfixauthdipany.commitd40cb9c.msi)

### Stress test parameters:
- Manager system: `centos`
- Launch CentOS agent: `true`
- Launch Ubuntu agent: `true`
- Launch Windows agent: `true`
- Test duration: `60`
- Scan frequency: `10`
- Modules to test: 
  - logcollector: `true`
  - syscheck: `true`
  - rootcheck: `true`
  - sca: `true`
  - osquery: `true`
  - syscollector: `true`
  - active-response: `true`
  - azure-logs: `true`
  - cis-cat: `true`
  - docker-listener: `true`
  - open-scap: `true`
  - vulnerability-detector: `false`
- FIM special test: `None`
- FIM sleep: `300`
- FIM max files: `25000`
- SCA sleep: `30`
- SCA max registers: `5000`
- Docker image name: `hello-world`
- Docker sleep: `5`
- Docker jobs: `10`
- Logcollector special test: `None`
- Logcollector max files: `50000`
- Logcollector creation sleep: `300`
- Logcollector write files: `1000`
- Logcollector write sleep: `10`
- Vuln sleep: `60`
- Authd port: `1515`
- Syscollector package: `imagemagick.app`
- Syscollector package version: `7.0.8.48`

Regards!.

**Comment by chemamartinez** [comment]
Hi @nicolaspapp, @DaveVG1,

The extra check added at the line https://github.com/wazuh/wazuh/pull/2533/files#diff-c1d16a58d2ed41485e829b7ff16b2ecfR935 is changing the expected behavior when calling `agent-auth -I any` by ignoring the option. Let see an example:

Agent CentOS is registered:
```
# agent-auth -A centos -m 192.168.0.100 -I any
```
Agent Debian is registered:
```
# agent-auth -A debian -m 192.168.0.100 -I any
```

**Expected `client.keys` in the manager:**
```
001 centos any {random_key}
002 debian any {random_key}
```

**Current `client.keys`:**
```
001 centos 192.168.0.122 {random_key}
002 debian 192.168.0.143 {random_key}
```

To sum up, `srcip` is never filled if `config.flags.use_source_ip`. Can you double-check it and revert that change if proceed?

Best regards.

File: src/os_auth/main-server.c [hunk 3]
```diff
@@ -927,10 +927,12 @@ void* run_dispatcher(__attribute__((unused)) void *arg) {
...
             {
-                sscanf(tmpstr," IP:\'%15[^\']\"",client_source_ip);
+                char format[15];
+                sprintf(format, " IP:\'%%%d[^\']\"", IPSIZE);
+                sscanf(tmpstr, format ,client_source_ip);
 
-                /* If IP: != 'src' overwrite the srcip */
-                if(strncmp(client_source_ip,"src",3) != 0)
+                /* If IP: != 'src' and 'use_source_ip="no" overwrite the srcip */
+                if((strncmp(client_source_ip,"src",3) != 0) && !config.flags.use_source_ip)
```
Comment (chemamartinez) [inline src/os_auth/main-server.c]: This line: https://github.com/wazuh/wazuh/pull/2533#issuecomment-575585523

**Comment by nicolaspapp** [comment]
Ok let´s see if I got it right now. Here is the **clients.key** according to both **-I** option an **use_source_ip** variable:

**Comment by nicolaspapp** [comment]
## Scenarios with USE_SOURCE_IP=0

### Scenario 1
- use_source_ip = 0
- commands:
```
agent-auth -A centos -I any -m 192.168.56.2
agent-auth -A debian -I any -m 192.168.56.2
```
- clients.key:
```
001 centos any {random_key}
002 debian any {random_key}
```

### Scenario 2
clients.key should be any if **use_soure_ip** is 0 and it the **-I** option is not used
- use_source_ip = 0
- commands:
```
agent-auth -A centos -m 192.168.56.2
agent-auth -A debian -m 192.168.56.2
```
- clients.key:
```
001 centos any {random_key}
002 debian any {random_key}
```

### Scenario 3
clients.key creates 2 hostname entries with same IP if **use_source_ip** is 0 and **-I** option is used
- use_source_ip = 0
- commands:
```
agent-auth -A centos -I 192.168.1.2 -m 192.168.56.2
agent-auth -A debian -I 192.168.1.2 -m 192.168.56.2
```
- clients.key:
```
001 centos 192.168.1.2 {random_key}
002 debian 192.168.1.2 {random_key}
```

**Comment by nicolaspapp** [comment]
## Scenarios with USE_SOURCE_IP=1

### Scenario 4
- use_source_ip = 1
- commands:
```
agent-auth -A centos -I any -m 192.168.56.2
agent-auth -A debian -I any -m 192.168.56.2
```
- clients.key:
```
{centos host was overriden by debian}
002 debian any {random_key}
```

### Scenario 5
- use_source_ip = 1
- commands:
```
agent-auth -A centos -m 192.168.56.2
agent-auth -A debian -m 192.168.56.2
```
- clients.key:
```
001 centos 192.168.56.3 {random_key}
002 debian 192.168.56.4 {random_key}
```

### Scenario 6
- use_source_ip = 1
- commands:
```
agent-auth -A centos -I 192.168.1.2 -m 192.168.56.2
agent-auth -A debian -I 192.168.1.2 -m 192.168.56.2
```
- clients.key:
```
{centos host was overriden by debian}
002 debian 192.168.1.2 {random_key}
```

**Comment by wazuhci** [comment]
The stress test has been launched. The charts will be posted when they are ready.
I will use the following configuration:
### General
- Commit: 1424a81
- Packages: 
  - [manager 3.12.0 0.branchfixauthdipany.commit1424a81 rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-manager-3.12.0-0.branchfixauthdipany.commit1424a81.x86_64.rpm)
  - [agent 3.12.0 0.branchfixauthdipany.commit1424a81 rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-agent-3.12.0-0.branchfixauthdipany.commit1424a81.x86_64.rpm)
  - [manager 3.12.0 0.branchfixauthdipany.commit1424a81 deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-manager_3.12.0-0.branchfixauthdipany.commit1424a81_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commit1424a81 deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-agent_3.12.0-0.branchfixauthdipany.commit1424a81_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commit1424a81 msi](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/windows/wazuh-agent-3.12.0-0.branchfixauthdipany.commit1424a81.msi)

### Stress test parameters:
- Manager system: `centos`
- Launch CentOS agent: `true`
- Launch Ubuntu agent: `true`
- Launch Windows agent: `true`
- Test duration: `60`
- Scan frequency: `10`
- Modules to test: 
  - logcollector: `true`
  - syscheck: `true`
  - rootcheck: `true`
  - sca: `true`
  - osquery: `true`
  - syscollector: `true`
  - active-response: `true`
  - azure-logs: `true`
  - cis-cat: `true`
  - docker-listener: `true`
  - open-scap: `true`
  - vulnerability-detector: `false`
- FIM special test: `None`
- FIM sleep: `300`
- FIM max files: `25000`
- SCA sleep: `30`
- SCA max registers: `5000`
- Docker image name: `hello-world`
- Docker sleep: `5`
- Docker jobs: `10`
- Logcollector special test: `None`
- Logcollector max files: `50000`
- Logcollector creation sleep: `300`
- Logcollector write files: `1000`
- Logcollector write sleep: `10`
- Vuln sleep: `60`
- Authd port: `1515`
- Syscollector package: `imagemagick.app`
- Syscollector package version: `7.0.8.48`

Regards!.

**Comment by nicolaspapp** [comment]
Changed the outcome for **Scenario 4.** If `any` is used, we will not verificate repeated even if the `use_source_ip` flag is set to yes

### Scenario 4
- use_source_ip = 1
- commands:
```
agent-auth -A centos -I any -m 192.168.56.2
agent-auth -A debian -I any -m 192.168.56.2
```
- clients.key:
```
001 centos any {random_key}
002 debian any {random_key}
```

**Comment by nicolaspapp** [comment]
This last change fixes scenario 3. It also simplifies the IP verification logic a lot. IP value (for both sourced or forced options) will be always verified unless it's value is `any `

### Scenario 3
clients.key creates 2 hostname entries with same IP if **use_source_ip** is 0 and **-I** option is used
- use_source_ip = 0
- commands:
```
agent-auth -A centos -I 192.168.1.2 -m 192.168.56.2
agent-auth -A debian -I 192.168.1.2 -m 192.168.56.2
```
- clients.key:
```
{centos host was overriden by debian}
002 debian 192.168.1.2 {random_key}
```

**Comment by wazuhci** [comment]
The stress test has been launched. The charts will be posted when they are ready.
I will use the following configuration:
### General
- Commit: 3263b4e
- Packages: 
  - [manager 3.12.0 0.branchfixauthdipany.commit3263b4e rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-manager-3.12.0-0.branchfixauthdipany.commit3263b4e.x86_64.rpm)
  - [agent 3.12.0 0.branchfixauthdipany.commit3263b4e rpm](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/rpm/var/wazuh-agent-3.12.0-0.branchfixauthdipany.commit3263b4e.x86_64.rpm)
  - [manager 3.12.0 0.branchfixauthdipany.commit3263b4e deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-manager_3.12.0-0.branchfixauthdipany.commit3263b4e_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commit3263b4e deb](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/deb/var/wazuh-agent_3.12.0-0.branchfixauthdipany.commit3263b4e_amd64.deb)
  - [agent 3.12.0 0.branchfixauthdipany.commit3263b4e msi](https://packages-dev.wazuh.com/warehouse/pullrequests/3.12/windows/wazuh-agent-3.12.0-0.branchfixauthdipany.commit3263b4e.msi)

### Stress test parameters:
- Manager system: `centos`
- Launch CentOS agent: `true`
- Launch Ubuntu agent: `true`
- Launch Windows agent: `true`
- Test duration: `60`
- Scan frequency: `10`
- Modules to test: 
  - logcollector: `true`
  - syscheck: `true`
  - rootcheck: `true`
  - sca: `true`
  - osquery: `true`
  - syscollector: `true`
  - active-response: `true`
  - azure-logs: `true`
  - cis-cat: `true`
  - docker-listener: `true`
  - open-scap: `true`
  - vulnerability-detector: `false`
- FIM special test: `None`
- FIM sleep: `300`
- FIM max files: `25000`
- SCA sleep: `30`
- SCA max registers: `5000`
- Docker image name: `hello-world`
- Docker sleep: `5`
- Docker jobs: `10`
- Logcollector special test: `None`
- Logcollector max files: `50000`
- Logcollector creation sleep: `300`
- Logcollector write files: `1000`
- Logcollector write sleep: `10`
- Vuln sleep: `60`
- Authd port: `1515`
- Syscollector package: `imagemagick.app`
- Syscollector package version: `7.0.8.48`

Regards!.

**Pull request merged**